File: swift/Day14.swift
import Foundation

struct CacheKey: Hashable {
    let polymer: String
    let step: Int
}

var cache: Dictionary<CacheKey, Dictionary<Character, UInt64>> = [:]

@main
struct Day14 {
    

    static func part2(input: Array<String>) -> UInt64 {
        let template: Array<Character> = input[0].map { value in value}
        
        var pairs: Dictionary<String, Character> = [:]
        input.suffix(input.count - 1).forEach { value in
            let line = value.components(separatedBy: " -> ")
//            let pair, insert = line
            pairs[line[0]] = line[1].first!
        }
        
        var count: Dictionary<Character, UInt64> = [template.last!: 1]
        for i in 0..<(template.count-1) {
            let sequenceCount = grow("\(template[i])\(template[i+1])", 40, pairs)
            count.merge(sequenceCount) { $0 + $1 }
        }
        
        let max: UInt64 = count.max { $0.value < $1.value }?.value ?? 0
        let min: UInt64 = count.min { $0.value < $1.value }?.value ?? 0
        return max - min
    }
    
    static func grow(_ polymer: String, _ step: Int, _ pairs: Dictionary<String, Character>) -> Dictionary<Character, UInt64> {
        guard step != 0 else { return [polymer[0]: 1] }
    
        if let cached = cache[CacheKey(polymer: polymer, step: step)] {
            return cached
        }
        
        if let insert = pairs[polymer] {
            let left = grow("\(polymer[0])\(insert)", step-1, pairs)
            let right = grow("\(insert)\(polymer[1])", step-1, pairs)
            let count = left.merging(right) { (old, new) in old + new}
            cache[CacheKey(polymer: polymer, step: step)] = count
            return count
        }
        return [polymer[0]: 1]
    }

    
    static func main() {
//        let testInput = Utils.readInput(name: "Day09_test")
//        assert(part1(input: testInput) == 1588)
//        assert(part2(input: testInput) == 1134)
        
        let input = Utils.readInput(name: "Day14")
//        print(part1(input: input), terminator: "\n")
        print(part2(input: input), terminator: "\n")
    }
}

File: kotlin/Day14.kt
@file:Suppress("PackageDirectoryMismatch", "PackageName")

package Day14

import readInput

fun part1(input: List<String>): Int {
    var template = mutableListOf<Char>()
    for (char in input[0]) {
        template.add(char)
    }

    val pairs = mutableMapOf<String, Char>()
    for (line in input.subList(2, input.size)) {
        val (pair, insert) = line.split(" -> ")
        pairs[pair] = insert.first()
    }

    for (step in 1..10) {
        val newPolymer = mutableListOf<Char>()
        template.windowed(2).forEach { chars ->
            val pair = chars.joinToString(separator = "")
            newPolymer.add(chars[0])
            pairs[pair]?.let { insert ->
                newPolymer.add(insert)
            }
        }
        newPolymer.add(template.last())
        template = newPolymer
    }
    val frequency = template.groupBy { it }.map { it.value.size }.sorted()
    return frequency.last() - frequency.first()
}


fun part2(input: List<String>): Long {
    val template = mutableListOf<Char>()
    for (char in input[0]) {
        template.add(char)
    }

    val pairs = mutableMapOf<String, Char>()
    for (line in input.subList(2, input.size)) {
        val (pair, insert) = line.split(" -> ")
        pairs[pair] = insert.first()
    }


    val count = template.windowed(2).fold(mapOf(template.last() to 1L)) { acc, list ->
        mergeMaps(acc, grow(list.joinToString(separator = ""), 40, pairs))
    }
    return count.maxOf { it.value } - count.minOf { it.value }
}

val cache = mutableMapOf<Pair<String, Int>, Map<Char, Long>>()

fun grow(polymer: String, step: Int, pairs: Map<String, Char>): Map<Char, Long> {
    if (step == 0) return mapOf(polymer[0] to 1L)

    cache[polymer to step]?.let { return it }

    pairs[polymer]?.let { insert ->
        val left = grow(listOf(polymer[0], insert).joinToString(separator = ""), step - 1, pairs)
        val right = grow(listOf(insert, polymer[1]).joinToString(separator = ""), step - 1, pairs)
        return mergeMaps(left, right).also { cache[polymer to step] = it }
    }
    return mapOf(polymer[0] to 1L)
}

fun mergeMaps(left: Map<Char, Long>, right: Map<Char, Long>): MutableMap<Char, Long> {
    val count = mutableMapOf<Char, Long>()
    left.forEach { (key, value) -> count[key] = value }
    right.forEach { (key, value) -> count[key] = count[key]?.plus(value) ?: value }
    return count
}

fun main() {

    // test if implementation meets criteria from the description, like:
    val testInput = readInput("Day14_test")
    check(part1(testInput) == 1588)
//    check(part2(testInput) == 2188189693529L)

    val input = readInput("Day14")
    println(part1(input))
    println(part2(input))
}
